py/server/tests/test_table.py

```diff
@@ -406,30 +423,66 @@ def test_agg_by_2(self):
             ["grp_id=(int)(i/5)", "var=(int)i", "weights=(double)1.0/(i+1)"]
         )
 
-        aggs = [
-            group(["aggGroup=var"]),
-            avg(["aggAvg=var"]),
-            count_("aggCount"),
-            partition("aggPartition"),
-            first(["aggFirst=var"]),
-            last(["aggLast=var"]),
-            max_(["aggMax=var"]),
-            median(["aggMed=var"]),
-            min_(["aggMin=var"]),
-            pct(0.20, ["aggPct=var"]),
-            std(["aggStd=var"]),
-            sum_(["aggSum=var"]),
-            abs_sum(["aggAbsSum=var"]),
-            var(["aggVar=var"]),
-            weighted_avg("var", ["weights"]),
-        ]
-
-        result_table = test_table.agg_by(aggs, ["grp_id"])
-        self.assertGreaterEqual(result_table.size, 1)
+        result_table = test_table.agg_by(self.aggs, ["grp_id"])
+        self.assertEqual(result_table.size, 2)
 
-        for agg in aggs:
+        for agg in self.aggs:
             result_table = test_table.agg_by(agg, "grp_id")
-            self.assertGreaterEqual(result_table.size, 1)
+            self.assertEqual(result_table.size, 2)
+
+    def test_agg_by_initial_groups_preserve_empty(self):
+        test_table = empty_table(10)
+        test_table = test_table.update(
+            ["grp_id=(int)(i/5)", "var=(int)i", "weights=(double)1.0/(i+1)"]
+        )
+
+        with self.subTest("no-initial-groups, no-by, preserve_empty only"):
+            t = test_table.where("grp_id > 2")
+            result_table = t.agg_by(self.aggs, preserve_empty=False)
+            self.assertEqual(result_table.size, 0)
+            result_table = t.agg_by(self.aggs, preserve_empty=True)
+            self.assertEqual(result_table.size, 1)
+            print(result_table.to_string())
+
+        with self.subTest("with initial-groups, no-by, and preserve_empty"):
+            init_groups = test_table.update("grp_id=i")
+            with self.assertRaises(DHError):
+                result_table = test_table.agg_by(self.aggs, initial_groups=init_groups)
```

Comment:
Is this raising an exception because not all groups are present?  If so, I don't think this concept is clear in the doc string above.

Reply:
I would comment this test.  I did read that in the docs and code.